crates/ironrdp-server/src/encoder/bitmap.rs

```diff
@@ -0,0 +1,111 @@
+use ironrdp_graphics::{
+    image_processing::PixelFormat,
+    rdp6::{ABgrChannels, ARgbChannels, BgrAChannels, BitmapStreamEncoder, RgbAChannels},
+};
+use ironrdp_pdu::{
+    bitmap::{self, BitmapData, BitmapUpdateData, Compression},
+    cursor::WriteCursor,
+    geometry::InclusiveRectangle,
+    PduEncode, PduError,
+};
+
+use crate::{BitmapUpdate, PixelOrder};
+
+// PERF: we could also remove the need for this buffer
+pub struct BitmapEncoder {
+    buffer: Vec<u8>,
+}
+
+impl BitmapEncoder {
+    pub fn new() -> Self {
+        Self {
+            buffer: vec![0; u16::MAX as usize],
+        }
+    }
+
+    pub fn encode(&mut self, bitmap: &BitmapUpdate, output: &mut [u8]) -> Result<usize, PduError> {
+        let row_len = bitmap.width * bitmap.format.bytes_per_pixel() as u32;
+        let chunk_height = u16::MAX as u32 / row_len;
+
+        let mut cursor = WriteCursor::new(output);
+        let chunks = bitmap.data.chunks((row_len * chunk_height) as usize);
+
+        let total = chunks.size_hint().0;
+        BitmapUpdateData::encode_header(total as u16, &mut cursor)?;
+
+        for (i, chunk) in chunks.enumerate() {
+            let height = chunk.len() as u32 / row_len;
+            let top = bitmap.top + i as u32 * chunk_height;
+
+            let encoder = BitmapStreamEncoder::new(bitmap.width as usize, height as usize);
+
+            let len = match bitmap.order {
+                PixelOrder::BottomToTop => {
+                    Self::encode_slice(encoder, bitmap.format, chunk, self.buffer.as_mut_slice())
+                }
+
+                PixelOrder::TopToBottom => {
+                    let bytes_per_pixel = bitmap.format.bytes_per_pixel() as usize;
+                    let pixels = chunk
+                        .chunks(row_len as usize)
+                        .rev()
+                        .flat_map(|row| row.chunks(bytes_per_pixel));
+
+                    Self::encode_iter(encoder, bitmap.format, pixels, self.buffer.as_mut_slice())
```

Comment:
Here we could have a problem - if the chunk size is bigger than the pre-allocated buffer, this code will throw an error (currently panic, as `ensure_size` is missing some places).

One of the possible solutions to this issue:
- We could have some method that will calculate the exact or at least the maximum amount of bytes required to encode the PDU. then we call it prior to the actual encoding step and resize the buffer accordingly.

@CBenoit what are your thoughts on this?

Reply:
Agreed!

That being said, currently it looks like `buffer` is always assigned a `Vec` with a capacity for `u16::MAX` elements (initially all filled with `0`s):
https://github.com/Devolutions/IronRDP/pull/167/files/37954589ab655b8a570d8de3872f2b1f789b78ed#diff-fb555a3444a021d9c141746de60538f9d8c408b6952f069372ab3b59d957fdfbR22
Is it normal? If we could find a smaller upper-bound this would help with memory usage I imagine

Reply:
I am looking for a way to remove the need for this buffer entirely. If I can't find a clean way to do it, I will make the buffer resizable.

Reply:
After looking into it some more, removing the need for this buffer would require some changes to the `BitmapData` (TS_BITMAP_DATA), splitting it into a header and data part similar to `BitmapStream`.

I would like to leave it like this for now and improve it in a later pr, if you are ok with it :)

Reply:
Sounds good to me!